dango/account/margin/src/execute.rs

```diff
@@ -47,3 +66,166 @@ pub fn receive(_ctx: MutableCtx) -> StdResult<Response> {
+    } = ctx
+        .querier
+        .query_health(ctx.contract, Some(ctx.funds.clone()))?;
+
+    // Ensure account is undercollateralized
+    ensure!(
+        utilization_rate > Udec128::ONE,
+        "account is not undercollateralized"
+    );
+
+    let health_factor = utilization_rate.checked_inv()?;
+    let target_health_factor = app_cfg.target_utilization_rate.checked_inv()?;
+    let liquidation_collateral_power = *app_cfg
+        .collateral_powers
+        .get(&liquidation_denom)
+        .ok_or_else(|| anyhow!("collateral power not found for chosen collateral"))?
+        .inner();
+
+    // Calculate liquidation bonus.
+    let bonus_cap = (total_adjusted_collateral_value
+        .checked_div(total_debt_value.checked_mul(liquidation_collateral_power)?))?
+    .saturating_sub(Udec128::ONE);
+    let liq_bonus = max(
+        *app_cfg.min_liquidation_bonus,
+        min(
+            bonus_cap,
+            min(*app_cfg.max_liquidation_bonus, Udec128::ONE - health_factor),
+        ),
+    );
+
+    // Calculate value of maximum repayable debt (MRD) to reach the target utilization rate.
+    // See derivation of the equation in [liquidation-math.md](book/notes/liquidation-math.md).
+    // It shouldn't be possible for the numerator to be negative, as the accunt should only be
+    // liquidatable if it is undercollateralized. If the denominator is negative (should only
+    // happen with an excessive minimum liquidation bonus), then the MRD is set to the account's
+    // total debt value.
+    let mrd_to_target_health = if target_health_factor
+        <= (Udec128::ONE + liq_bonus) * liquidation_collateral_power
+    {
+        total_debt_value
+    } else {
+        let numerator = total_debt_value
+            .checked_mul(target_health_factor)?
+            .checked_sub(total_adjusted_collateral_value)?;
+        let denominator = target_health_factor
+            .checked_sub((Udec128::ONE + liq_bonus).checked_mul(liquidation_collateral_power)?)?;
+        numerator.checked_div(denominator)?
+    };
+
+    // Calculate the maximum debt that can be repaid based on the balance of the
+    // chosen collateral.
+    let collateral_price = ctx
+        .querier
+        .query_price(app_cfg.addresses.oracle, &liquidation_denom)?;
+    let liquidation_collateral_value =
+        collateral_price.value_of_unit_amount(collaterals.amount_of(&liquidation_denom))?;
+    let mrd_from_chosen_collateral =
+        liquidation_collateral_value.checked_div(Udec128::ONE + liq_bonus)?;
+
+    // Calculate the debt value to repay.
+    let debt_repay_value = [
+        mrd_to_target_health,
+        mrd_from_chosen_collateral,
+        total_debt_value,
+    ]
+    .into_iter()
+    .min()
+    .ok_or_else(|| anyhow!("unable to calculate debt repay value"))?;
+
+    if debt_repay_value.is_zero() {
+        bail!(
+            "Debt repay value is zero. Probably the account either has no debt, or no collateral."
+        );
+    }
```

Comment:
I like to use `ensure!` over `bail!` whenever possible:

```rust
ensure!(
    debt_repay_value.is_non_zero(),
    "..."
)
```

Also, don't uppercase the error message. See: https://github.com/rust-lang/api-guidelines/issues/79